src/Hyperion/Extensions/StreamEx.cs

```diff
@@ -190,9 +189,37 @@ public static string ReadString(this Stream stream, DeserializerSession session)
 
             var buffer = session.GetBuffer(length);
 
-            stream.Read(buffer, 0, length);
+            stream.ReadFull(buffer, 0, length);
             var res = StringEx.FromUtf8Bytes(buffer, 0, length);
             return res;
         }
+
+        /// <summary>
+        /// Repeats reading from stream until requested bytes were read.
+        /// Returns with partial result if stream can't provide enough bytes
+        /// Fixes issue: https://github.com/akkadotnet/Hyperion/issues/40
+        /// Reference for allowed partial streams: https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?redirectedfrom=MSDN&view=netcore-3.1#System_IO_Stream_Read_System_Byte___System_Int32_System_Int32_
+        /// -> "An implementation is free to return fewer bytes than requested even if the end of the stream has not been reached."
+        /// </summary>
+        public static int ReadFull(this Stream stream, byte[] buffer, int offset, int count)
+        {
+            // fast path for streams which doesn't deliver partial results
+            var totalReadBytes = stream.Read(buffer, offset, count);
+            if (totalReadBytes == count)
+                return totalReadBytes;
+
+            // support streams with partial results
+            do
+            {
+                var readBytes = stream.Read(buffer, offset + totalReadBytes, count - totalReadBytes);
+                if (readBytes == 0)
+                    break;
```

Comment:
I'm worried about this line.
In Microsoft implementation for BinaryReader, a primitive read that fails to read the full requested byte count throws an exception instead of stopping read, this could possibly turn into a silent error later on.

Reply:
@Arkatufus I agree. So we should check here if the required bytes were read and if not throw an appropriate exception

Reply:
By just returning if no bytes were read the behavior should be the same as without this PR. The exception will then happen somewhere else in Hyperion which assumes that all bytes were read.

Reply:
Ok, can you make this change then @Ralf1108 ?

Reply:
Change this to:
```C#
if(readBytes == 0 && !allowPartial)
    throw new EndOfStreamException($"Unexpected EOF. Expected: {count + totalReadBytes}, Received: {totalReadBytes}");
else
    break;
```